### ISSUE TITLE
Using java.util.randomUUID for setInsertId and no new data after some time

### ISSUE THREAD

**Issue by GoogleCodeExporter** [runt18/google-bigquery#172]
```
What steps will reproduce the problem?
1. Have a queue of warmed up bq connections. Once bq insertall is done, put the 
connection back to queue
2. Keep populating and sending batch of rows with insertid that is set to 
java.util.randomUUID

What is the expected output? What do you see instead?
expect to see new rows are coming to bq table but though there is no failure

What version of the product are you using? On what operating system?
OS is ubuntu
google-api-services-bigquery:v2-rev159-1.19.0
google-api-client:1.19.0
google-http-client-jackson2:1.19.0

Please provide any additional information below.


```

Original issue reported on code.google.com by `bnemu...@zulily.com` on 22 Oct 2014 at 4:30

**Comment by GoogleCodeExporter** [comment]
```
The details for the coding can be found at 
http://stackoverflow.com/questions/26479462


```

Original comment by `bnemu...@zulily.com` on 22 Oct 2014 at 4:49

**Comment by GoogleCodeExporter** [comment]
```
Can you elaborate on what you mean by "warmed up" connections?

From the perspective of insertAll, the warm up period occurs 1-2 minutes after 
the first successful insertion request.

Can you also provide the project id and dataset id?
```

Original comment by `seanc...@google.com` on 23 Oct 2014 at 8:28

**Comment by GoogleCodeExporter** [comment]
```
Like I said on the bug, after instantiating bq connection and push the
first row in, we put that connection to a circler queue to be reused for
another batch. The queue size can go up to a given number of connections
and now it is set to 100. This way, we can minimize the warm up period for
any subsequent streams.

It is working in production and had streamed more than billions of rows
already. After implemented this retry logic with insertid to improve the
data quality, first few minutes it is fine then no failure nor new record
coming in.


ProjectId:test-altar-543
Tables:zip_integration.zip_SignIn or zip_integration.zip_SignUp
I had to rollback on tuesday around 6PM, so the issue was there between
3PM-6PM.

I will go ahead and test more today and over the weekend to reproduce the
issue!

thanks for the response!
```

Original comment by `bnemu...@zulily.com` on 23 Oct 2014 at 8:43

**Comment by GoogleCodeExporter** [comment]
```
Are your times in PST, or something else?

Between ~3PM and ~5:30PM our system show ~150k rows accepted during that time 
frame. That looks roughly in line with the previous days numbers.

1) Is that inline with your expectations.
2) Did you see the rows appears as expected later, or never see rows appear at 
all?
```

Original comment by `seanc...@google.com` on 23 Oct 2014 at 9:43

**Comment by GoogleCodeExporter** [comment]
```
We are in PST. I am setting up the test in isolated environment and I'll pump a 
lot of data in the same project against sandbox.zip_Visit. 

During that time period, I did several iterations rollback and redeploy to 
unblock the team so the data partially coming. Eventually, I gave up as I 
noticed that rollback up was helping introducing new rows.

Stay Tuned!

```

Original comment by `bnemu...@zulily.com` on 23 Oct 2014 at 10:14

**Comment by GoogleCodeExporter** [comment]
```
Deploy the retry logic with insertid and bounced the service
Between 9PM PST and 10PM on 10/27/2014, Pushed 100k rows  each time to 
zip_integration.zip_Visit on that same project

First attempt resulted in ~50% data lost and the second one is like 75% then no 
more data in bq
100000 records submitted in 135020 milliseconds and bq has 56604 rows
100000 records submitted in 135124 milliseconds and bq has 69409 rows
100000 records submitted in 133075 milliseconds and bq has 80449 rows
100000 records submitted in 135792 milliseconds and bq has 80449 rows
100000 records submitted in 132778 milliseconds and bq has 80449 rows
100000 records submitted in 137151 milliseconds and bq has 80449 rows

Without, retry logic and insertid set to uuid, the load loss is usually 0.1%
I can repro the same If I bounce the service that maintains the queue of 
connections.

```

Original comment by `bnemu...@zulily.com` on 28 Oct 2014 at 5:10

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
I rolled back the retry logic and kept only code change "new 
Rows().setJson(this.row).setInsertId(UUID.randomUUID().toString()) "
Since ~9PMThru multiple iterations, I tried to push ~4 million but only 2.8 
million rows in bq table zip_integration.zip_SignIn
Now no more rows are going to the table regardless the size of the batch!

When I exclude setinsertid with UUID, all is working fine as that's what it is 
in production.

I'd truly appreciate any help!



```

Original comment by `bnemu...@zulily.com` on 29 Oct 2014 at 6:36

**Comment by GoogleCodeExporter** [comment]
```
Our system only registered ~2.856 M deduplicated rows, and after deduplication 
saw 2.851M rows, meaning we only considered ~5k of your  insertions as 
duplicates.

Can you provide the most updated code sample, including your insertion id 
setting logic?

Also, what are you using to estimate the ~4M rows expectation you mentioned? 
```

Original comment by `seanc...@google.com` on 29 Oct 2014 at 7:34

**Comment by GoogleCodeExporter** [comment]
```
Additionally, please provide your request shape...

i.e. Does your system continuously fire insertion requests in parallel, or does 
it poll some queue on some fixed interval? How many concurrent requests will 
your system generate, of what batching size, etc.

The more information the better in determining where the issue may be. Also, 
what are the error statistics counters reporting? All 0? If so, have you 
synthetically incremented them to make sure they're reporting correctly?
```

Original comment by `seanc...@google.com` on 29 Oct 2014 at 7:55

**Comment by GoogleCodeExporter** [comment]
```
It is mimicking jdbc for big query so we have those objects as 
PreparedStatement. BQConnection etc.
For each row, the application populates TableRow object and calls addBatch() 
method;
  @Override
  public void addBatch() throws SQLException {
    listRows.add(new Rows().setJson(this.row).setInsertId(UUID.randomUUID().toString()));
    this.row = new TableRow();
  }

Once there is a batch of 500 rows in a queue, it calls executeBatch() for that 
queue and we can have multiple queues that can be pushed in parallel. On top of 
that, in production, we are running the service as cluster.


  @Override
  public boolean InsertBQTable(String DataSetId, String TableId, List<Rows> rowList) throws GeneralSecurityException,
      IOException {

    TableDataInsertAllRequest content = new TableDataInsertAllRequest().setRows(rowList);
    TableDataInsertAllResponse response = bq.tabledata().insertAll(bqProjectId, DataSetId, TableId, content).execute();
    if (response.getInsertErrors()!=null)
      warn("Inserting One of the rows has failed no retry");
    return true;
  }  

As you see, I don't have any retry logic and I checked unique value is assigned 
to insertid property of each row.
Last night after posting the previous message, I left pushing 10m rows and 3 
times 100k push this morning. Nothing new appeared in that bq 
table;zip_integration.zip_SignIn ???

Please note that we maintain queue of connection and adjustable connection 
queue size is 100 for now. If recycle the service which will release all those 
connections, the issue gets resolved for sometime. In production, we don't have 
such issue as the rows showing up as expected and only code change here ; 
.setInsertId(UUID.randomUUID().toString())


Thanks for the help!








```

Original comment by `bnemu...@zulily.com` on 29 Oct 2014 at 5:19

**Comment by GoogleCodeExporter** [comment]
```
I think there may be something wrong on your connection end. Here are our 
recorded *request* counts for all tables in your project during the time period 
in question (each request has multiple rows):

+---------------+----------+
| time          | COUNT(*) |
+---------------+----------+
| 2014-10-28 07 |        5 |
| 2014-10-28 08 |        8 |
| 2014-10-28 09 |        5 |
| 2014-10-28 10 |      247 |
| 2014-10-28 11 |     3689 |
| 2014-10-28 12 |     3634 |
| 2014-10-28 13 |     4083 |
| 2014-10-28 14 |     1122 |
| 2014-10-28 15 |     3597 |
| 2014-10-28 16 |    15732 |
| 2014-10-28 20 |      656 |
| 2014-10-28 21 |     3364 |
| 2014-10-28 22 |    29357 |
| 2014-10-28 23 |     1367 |
| 2014-10-29 11 |      409 |
| 2014-10-29 13 |      232 |
| 2014-10-29 14 |      129 |
| 2014-10-29 15 |       52 |
| 2014-10-29 16 |       18 |
| 2014-10-29 17 |        5 |
| 2014-10-29 23 |        1 |
| 2014-10-30 09 |        1 |
| 2014-10-30 10 |      100 |
| 2014-10-30 11 |       69 |
| 2014-10-30 12 |       56 |
| 2014-10-30 13 |       17 |
| 2014-10-30 14 |       63 |
| 2014-10-30 15 |        6 |
| 2014-10-30 16 |        1 |
| 2014-10-30 17 |        9 |
| 2014-10-30 18 |        9 |
| 2014-10-30 23 |       48 |
| 2014-10-31 09 |        1 |
| 2014-10-31 10 |       17 |
| 2014-10-31 11 |       26 |
| 2014-10-31 12 |        5 |
| 2014-10-31 13 |        7 |
| 2014-10-31 14 |       38 |
| 2014-10-31 15 |       28 |
| 2014-10-31 16 |        2 |
| 2014-10-31 17 |        1 |
| 2014-11-02 07 |        1 |
| 2014-11-02 13 |        1 |
| 2014-11-02 16 |        1 |
| 2014-11-03 09 |        8 |

+---------------+----------+

You can see your test load from 10/28, 8pm - 12pm or so, but nothing 
significant after that. This logging is performed at the edges of our system, 
before the point before any row level considerations are made. We also don't 
see significant error responses etc (i.e. all of these appear to be successful 
insertion requests).

Is it possible for you to add detailed logging around your request/responses to 
see that they're actually making it to our network?

```

Original comment by `seanc...@google.com` on 3 Nov 2014 at 5:47

**Comment by GoogleCodeExporter** [comment]
```
Sorry.. that should read "10/28 8pm to 10/29 12am"
```

Original comment by `seanc...@google.com` on 3 Nov 2014 at 5:50

**Comment by GoogleCodeExporter** [comment]
```
Does this continue to be an issue?
```

Original comment by `seanc...@google.com` on 5 Dec 2014 at 8:03

**Comment by GoogleCodeExporter** [comment]
```
Sorry, I gave up testing after some point and had to accept 0.01% failure rate 
for now.

```

Original comment by `bnemu...@zulily.com` on 5 Jan 2015 at 9:48

**Comment by GoogleCodeExporter** [comment]
```
Just to clarify, The original bug text indicates that rows stop getting 
accepted (not a 0.01% failure rate). Is that still the case?
```

Original comment by `seanc...@google.com` on 12 Jan 2015 at 5:34

**Comment by GoogleCodeExporter** [comment]
```
To improve that failure rate of 0.01, I introduced the retry logic with GUID on 
row level to rely on the dedup logic on your side. Once it is utilized, I was 
facing the issue I described above.


```

Original comment by `bnemu...@zulily.com` on 12 Jan 2015 at 6:19

**Comment by GoogleCodeExporter** [comment]
```
Hrm. Ok... Well if this isn't causing issues for you then I'll close this out. 
Please reopen if you need this behavior and run into this issue again.
```

Original comment by `seanc...@google.com` on 12 Jan 2015 at 7:02
- Changed state: **WontFix**
